File: example-code/prove.py
import json
import sys
import time
import requests  # pip install requests

# NOTE: Provide your API key here
api_key = ""

# NOTE: Provide your circuit ID and the path to your circuit input here
CIRCUIT_ID = ""
PROOF_INPUT = "../gnark/cubic/input.json"

API_VERSION = "v1"
API_URL = f"https://forge.sindri.app/api/{API_VERSION}/"
api_key_querystring = f"?api_key={api_key}"
headers_json = {"Accept": "application/json"}
headers_multipart = {"Accept": "multipart/form-data"}
headers_urlencode = {
    "Accept": "application/json",
    "Content-Type": "application/x-www-form-urlencoded",
}

# Initiate proof generation 3 times
for _ in range(3):
  print("Proving circuit...")
  with open(PROOF_INPUT, "r") as proof_file:
    proof_input = json.dumps(json.load(proof_file))

  response = requests.post(
      API_URL + f"circuit/{CIRCUIT_ID}/prove" + api_key_querystring,
      headers=headers_urlencode,
      data={
          "proof_input": proof_input
          # "prover_implementation": json.dumps({"name":"CPU Stock 1"})
      },
  )
  assert response.status_code == 201, f"Expected status code 201, received {response.status_code}."
  proof_id = response.json()["proof_id"]  # Obtain proof_id

  # Poll proof detail until it has a status of Ready or Failed
  TIMEOUT = 1200  # timeout after 20 minutes
  for i in range(TIMEOUT):
    response = requests.get(
        API_URL + f"proof/{proof_id}/detail" + api_key_querystring,
        headers=headers_json,
        params={
            "include_proof_input": False,
            "include_public": True,
            "include_verification_key": False,
            "include_proof": False,
        },
    )
    assert (response.status_code == 200
            ), f"Expected status code 200, received {response.status_code}."
    status = response.json()["status"]
    if status in ["Ready", "Failed"]:
      print(f"Proof poll exited after {i} seconds with status: {status}")
      break
    time.sleep(1)
  else:
    sys.exit("Proof polling timed out")

  # Check for proving issues
  if status == "Failed":
    sys.exit("Proving failed")
  else:
    # Retrieve output from the proof
    public_output = response.json()["public"]
    print("Proof generation succeeded!")
    print(json.dumps(response.json(), indent=2))

File: circuit_tutorials/halo2/axiom-v0.3.0/float_radius/zk_execute/src/bin/prove.rs
use zk_execute::{
    headers_json,
    poll_status
};
use std::{
    env,
    fs,
    io::{BufWriter,Read,Write}
};
use reqwest::Client;
use serde_json::Value;


#[tokio::main]
async fn main() {

    let api_key: String = env::var("SINDRI_API_KEY").unwrap();
    let api_url_prefix: &str = "https://sindri.app/api/";
    
    let api_version: &str = "v1/";
    let api_url: String = api_url_prefix.to_owned()  + api_version;

    println!("Reading circuit details locally");
    let mut file = fs::File::open("./data/compile_out.json").unwrap();
    let mut data = String::new();
    file.read_to_string(&mut data).unwrap();
    let circuit_data: Value = serde_json::from_str(&data).unwrap();
    let circuit_id = &circuit_data["circuit_id"].as_str().unwrap();

    // Initiate proof generation.
    println!("Reading proof input from example-input.json");
    let mut proof_input_file = fs::File::open("example-input.json").unwrap();
    let mut proof_input = String::new();
    proof_input_file.read_to_string(&mut proof_input).unwrap();
    let map = serde_json::json!({"proof_input": proof_input});

    println!("Requesting a proof");
    let client = Client::new();
    let response = client
        .post(format!("{api_url}circuit/{circuit_id}/prove"))
        .headers(headers_json(&api_key))
        .json(&map)
        .send()
        .await
        .unwrap();
    assert_eq!(&response.status().as_u16(), &201u16, "Expected status code 201");
    let response_body = response.json::<Value>().await.unwrap();
    let proof_id = response_body["proof_id"].as_str().unwrap();

    // Poll proof detail until it has a status of Ready or Failed
    let proof_data = poll_status(
        format!("proof/{proof_id}/detail"),
        &api_url,
        &api_key,
        600).await;
    if proof_data["status"].as_str().unwrap().contains("Failed") {
        println!("Proving failed.");
        std::process::exit(1);
    }

    println!("Saving proof details locally");
    let file = fs::File::create("./data/prove_out.json").unwrap();
    let mut writer = BufWriter::new(file);
    serde_json::to_writer_pretty(&mut writer, &proof_data).unwrap();
    writer.flush().unwrap(); 

}
